cpp/src/transform/arrow_to_cudf.cu

```diff
@@ -0,0 +1,263 @@
+    const std::shared_ptr<arrow::Array> arrow_col,
+    data_type type,
+    bool skip_mask,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto data_buffer = arrow_col->data()->buffers[1];
+    auto num_rows    = data_buffer->size() / sizeof(T);
+    auto has_nulls   = skip_mask ? false : arrow_col->null_bitmap_data() != nullptr;
+    auto col         = make_fixed_width_column(
+      type, num_rows, has_nulls ? mask_state::UNINITIALIZED : mask_state::UNALLOCATED, stream, mr);
+    auto mutable_column_view = col->mutable_view();
+    cudaMemcpy(mutable_column_view.data<void*>(),
+               data_buffer->data(),
+               data_buffer->size(),
+               cudaMemcpyHostToDevice);
+    if (has_nulls) {
+      cudaMemcpy(mutable_column_view.null_mask(),
+                 arrow_col->null_bitmap_data(),
+                 arrow_col->null_bitmap()->size(),
+                 cudaMemcpyHostToDevice);
+    }
+
+    return std::make_pair<std::unique_ptr<column>, column_view>(
+      std::move(col),
+      cudf::detail::slice(col->view(),
+                          static_cast<size_type>(arrow_col->offset()),
+                          static_cast<size_type>(arrow_col->length())));
+  }
+
+  template <typename T>
+  std::enable_if_t<is_compound<T>(), std::pair<std::unique_ptr<column>, column_view>> operator()(
+    const std::shared_ptr<arrow::Array> arrow_col,
+    data_type type,
+    bool skip_mask,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    std::unique_ptr<rmm::device_buffer> mask = std::make_unique<rmm::device_buffer>(0);
+    std::unique_ptr<column> out_col;
+
+    if (arrow_col->null_bitmap_data() != nullptr) {
+      mask = std::make_unique<rmm::device_buffer>(arrow_col->null_bitmap()->size(), stream, mr);
+      cudaMemcpy(mask->data(),
+                 arrow_col->null_bitmap_data(),
+                 arrow_col->null_bitmap()->size(),
+                 cudaMemcpyHostToDevice);
+    }
+    std::vector<std::unique_ptr<column>> child_column;
+
+    if (std::is_same<T, cudf::string_view>::value) {
+      auto str_array    = static_cast<arrow::StringArray*>(arrow_col.get());
+      auto offset_array = std::make_shared<arrow::Int32Array>(
+        str_array->value_offsets()->size() / sizeof(int32_t), str_array->value_offsets(), nullptr);
+      auto char_array = std::make_shared<arrow::Int8Array>(
+        str_array->value_data()->size(), str_array->value_data(), nullptr);
+
+      auto offsets_column = std::move(type_dispatcher(data_type(type_id::INT32),
+                                                      dispatch_to_cudf_col{},
+                                                      offset_array,
+                                                      data_type(type_id::INT32),
+                                                      true,
+                                                      mr,
+                                                      stream)
```

Comment:
Seems a waste to call the `type_dispatcher` for a hardcoded type. Can the functor just be called directly?
```
auto offsets_column = std::move(
dispatch_to_cudf_col{}.operator()<int32_t>(offset_array,
                                           data_type(type_id::INT32), true, mr, stream)
.first);
```

Reply:
That's a nice option.